allennlp/models/semantic_parsing/atis/atis_semantic_parser.py

```diff
@@ -0,0 +1,592 @@
+        sql_queries : List[str], otpional (default=None)
+            A list of the SQL queries that are given during training or validation.
+        """
+        initial_info = self._get_initial_state_and_scores(utterance, world, actions, linking_scores)
+        initial_state = initial_info["initial_state"]
+        batch_size = list(utterance.values())[0].size(0)
+
+        if target_action_sequence is not None:
+            # Remove the trailing dimension (from ListField[ListField[IndexField]]).
+            target_action_sequence = target_action_sequence.squeeze(-1)
+            target_mask = target_action_sequence != self._action_padding_index
+        else:
+            target_mask = None
+
+        if self.training:
+            return self._decoder_trainer.decode(initial_state,
+                                                self._decoder_step,
+                                                (target_action_sequence, target_mask))
+        else:
+            # TODO(kevin) Move some of this functionality to a separate method for computing validation outputs.
+            action_mapping = {}
+            for batch_index, batch_actions in enumerate(actions):
+                for action_index, action in enumerate(batch_actions):
+                    action_mapping[(batch_index, action_index)] = action[0]
+            outputs: Dict[str, Any] = {'action_mapping': action_mapping}
+            outputs['linking_scores'] = linking_scores
+            if target_action_sequence is not None:
+                outputs['loss'] = self._decoder_trainer.decode(initial_state,
+                                                               self._decoder_step,
+                                                               (target_action_sequence, target_mask))['loss']
+            num_steps = self._max_decoding_steps
+            # This tells the state to start keeping track of debug info, which we'll pass along in
+            # our output dictionary.
+            initial_state.debug_info = [[] for _ in range(batch_size)]
+            best_final_states = self._beam_search.search(num_steps,
+                                                         initial_state,
+                                                         self._decoder_step,
+                                                         keep_final_unfinished_states=False)
+            outputs['best_action_sequence'] = []
+            outputs['debug_info'] = []
+            outputs['entities'] = []
+            outputs['logical_form'] = []
+            outputs['sql_queries'] = []
+            outputs['utterance'] = []
+            outputs['tokenized_utterance'] = []
+
+            for i in range(batch_size):
+                # Decoding may not have terminated with any completed logical forms, if `num_steps`
+                # isn't long enough (or if the model is not trained enough and gets into an
+                # infinite action loop).
+                if i in best_final_states:
```

Comment:
It's probably a bit nicer to read if you just do

```python
if i not in best_final_states:
    continue
```

Then there's less indentation on the rest.  We should probably change our other models to do this, too. (Not saying that you should do that in this PR.)

Reply:
Though, you probably want to `continue` only after you've counted some zeros for your metrics, so you don't overestimate your accuracy.